usr/sbin/pkcstok_migrate/pkcstok_migrate.c

```diff
@@ -0,0 +1,2367 @@
+    if (rc != CKR_OK) {
+        TRACE_ERROR("aes_256_unwrap failed with rc=%08X.\n", rc);
+        goto done;
+    }
+
+    rc = CKR_OK;
+
+done:
+
+    if (fp)
+        fclose(fp);
+
+    return rc;
+}
+
+/**
+ * loads the old des3 masterkey from given file with given PIN.
+ */
+CK_RV load_masterkey_00(const char *mkfile, const char *pin,
+                        CK_BYTE *masterkey)
+{
+    CK_BYTE des3_key[3 * DES_KEY_SIZE];
+    char hash_sha[SHA1_HASH_SIZE];
+    char pin_md5_hash[MD5_HASH_SIZE];
+    unsigned char *cipher = NULL;
+    unsigned char *clear = NULL;
+    unsigned long cipher_len, clear_len;
+    CK_RV ret;
+    int rc;
+    FILE *fp = NULL;
+
+    clear_len = cipher_len =
+        (MASTER_KEY_SIZE + SHA1_HASH_SIZE +
+        (DES_BLOCK_SIZE - 1)) & ~(DES_BLOCK_SIZE - 1);
+
+    fp = fopen(mkfile, "r");
+    if (!fp) {
+        TRACE_ERROR("Cannot open %s, errno=%s\n", mkfile, strerror(errno));
+        return CKR_FUNCTION_FAILED;
+    }
+
+    cipher = malloc(cipher_len);
+    clear = malloc(clear_len);
+    if (cipher == NULL || clear == NULL) {
+        ret = CKR_HOST_MEMORY;
+        goto done;
+    }
+
+    rc = fread(cipher, cipher_len, 1, fp);
+    if (rc != 1) {
+        TRACE_ERROR("Cannot read %ld bytes from %s\n", cipher_len, mkfile);
+        ret = CKR_FUNCTION_FAILED;
+        goto done;
+    }
+
+    /* decrypt the masterkey */
+    ret = compute_md5((char *)pin, strlen(pin), pin_md5_hash);
+    if (ret) {
+        TRACE_ERROR("Error calculating MD5 of PIN, ret=%08lX\n",ret);
+        goto done;
+    }
+
+    memcpy(des3_key, pin_md5_hash, MD5_HASH_SIZE);
+    memcpy(des3_key + MD5_HASH_SIZE, pin_md5_hash, DES_KEY_SIZE);
+
+    ret = sw_des3_cbc_decrypt(cipher, cipher_len, clear,
+                              &clear_len, (unsigned char *) "12345678",
+                              des3_key);
+    if (ret != CKR_OK) {
+        TRACE_ERROR("failed to decrypt master key file after read, ret=%08lX\n",ret);
+        ret = CKR_FUNCTION_FAILED;
+        goto done;
+    }
+
+    /* compare the hashes to verify integrity */
+    ret = compute_sha1((char *)clear, MASTER_KEY_SIZE, hash_sha);
+    if (ret) {
+        TRACE_ERROR("Failed to compute sha1 for masterkey, ret=%08lX\n",ret);
+        goto done;
+    }
+
+    if (memcmp(hash_sha, clear + MASTER_KEY_SIZE, SHA1_HASH_SIZE) != 0) {
+        TRACE_ERROR("%s appears to be tampered! Cannot migrate.\n", mkfile);
+        ret = CKR_FUNCTION_FAILED;
+        goto done;
+    }
+
+    memcpy(masterkey, clear, MASTER_KEY_SIZE);
+    ret = 0;
+
+done:
+    if (fp)
+        fclose(fp);
+    free(clear);
+    free(cipher);
+
+    return ret;
+}
+
+/**
+ * gets the stdll name from the given opencryptoki.conf file.
+ */
+CK_RV get_stdll_name(FILE *fp, char **stdll_name)
+{
+    char line[80], parm[80];
+    char *tmp_name;
+
+    snprintf(parm, sizeof(parm), "stdll =");
+    while (fgets(line, sizeof(line), fp)) {
+        if (strstr(line, "stdll =")) {
```

Comment:
I would not assume that there is always one blank between 'stdll' and '='....

Reply:
ok